rfcs/0074-tvmscript-unified-printer.md

```diff
@@ -0,0 +1,475 @@
+def func_a(A: T.Buffer[(1,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+
+@T.prim_func
+def func_b(A: T.Buffer[(8,), "int32"]) -> None:
+                       ^^^^
+    A[0] = 0
+```
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This section introduces the design philosophy of the printer, and demonstrates
+the proposed user-facing APIs where users means IR developers.
+
+## Two-Stage Translation
+
+Traditionally in TVM stack, printing is a single-stage process. The printer
+assumes certain syntax of the target language, and therefore, so far there are
+3 different printers all for TIR: ReprPrinter, TIRTextPrinter,
+TVMScriptPrinter.
+
+We extend the idea of the existing Doc class at
+[src/printer/doc.h#L67](https://github.com/apache/tvm/blob/main/src/printer/doc.h#L67)
+to allow better consistency and scalability. An IR, which could
+be TIR, Relax or any other ones developed by third-party vendors, is first
+translated to an intermediate Doc node tree, and then the Doc tree is mapped to a target
+language, for example, Python, C++ IRBuilder API, or Rust.
+
+**Stage 1 [TVM IR => Doc]**. On the first stage, the printer needs to take care
+of translating a TVM IR to Doc tree. As an example, `tir.For` is translated to
+`ForDoc` without having to worry about the underlying language. Note that some
+complicated nodes in TVM IR, for example, `PrimFunc`, could be translated to
+multiple IR elements, including `FunctionDoc` and a few `StmtDoc`.
+
+During the translation from IR to Doc tree, it is possible that some statement
+influences the syntax of its children or vice verse, especially for syntactic
+sugars and declaring undefined variables in IR fragment printing. Therefore, a
+generic data structure `Frame` is introduced to allow retrieval and
+manipulation the relevant context information.
+
+**Stage 2. [Doc => target language]**. On the second stage, Doc tree is then
+honestly translated to the target language in text format. For example, when
+the target language is python, `ForDoc` is translated to python’s for loop
+syntax:
+
+```python
+for ... in ...:
+  ...
+```
+
+When the target language becomes python IRBuilder, `ForDoc` is translated to:
+
+```cpp
+with T.For(...):
+  ...
+```
+
+For generality, the Doc tree is designed to select minimal elements that exist
+in languages used in developing TVM. A full spec of the Doc could be found in
+the next section.
+
+## Distributed Registration
+
+As a major engineering challenge for TVMScript to scale to multiple IRs, the
+existing printing logic has to be engineered, maintained and re-engineered in a
+single file, which has brought significant confusion for developing multi-level
+IRs for TVM Unity.
+
+Inspired by the pass infrastructure, as well as the ReprPrinter in TVM, we
+propose to develop the infrastructure to enable distributed registration, and
+further allows printer for different levels of IR to be registered in separate
```

Comment:
I'm not quite sure if I understand this correctly. Can you give me any example?

Reply:
This means the printer for TIR and Relax can be defined in different `cc` files and doesn't know the existence of each other. For example, the folder structure of TIR printer and Relax printer will be like
```bash
src/script/printer/
├── core # Core infra, which is IR-agnostic
├── tir # TIR dialect 
│   ├── expr.cc
│   └── ...
└── relax # Hypothetical Relax dialect (not part of our RFC)
    └── ...
```

An implication is that printer for an IR can be loaded as dynamic library, which is a building block to support third-party IR.

Reply:
Interesting. Thank you for the clarification! Cool work!